lean4-infoview/src/infoview/tooltips.tsx

```diff
@@ -0,0 +1,221 @@
+
+export const Tooltip = forwardAndUseRef<HTMLDivElement,
+  React.HTMLProps<HTMLDivElement> &
+    { reference: HTMLElement | null,
+      tooltipContent: TooltipContent,
+      placement?: Popper.Placement,
+      onFirstUpdate?: (_: Partial<Popper.State>) => void
+    }>((props_, divRef, setDivRef) => {
+  const {reference, tooltipContent, placement: preferPlacement, onFirstUpdate, ...props} = props_
+
+  // We remember the global trend in placement (as `globalPlacement`) so tooltip chains can bounce
+  // off the top and continue downwards or vice versa and initialize to that, but then update
+  // the trend (as `ourPlacement`).
+  const globalPlacement = React.useContext(TooltipPlacementContext)
+  const placement = preferPlacement ? preferPlacement : globalPlacement
+  const [ourPlacement, setOurPlacement] = React.useState<Popper.Placement>(placement)
+
+  // https://popper.js.org/react-popper/v2/faq/#why-i-get-render-loop-whenever-i-put-a-function-inside-the-popper-configuration
+  const onFirstUpdate_ = React.useCallback((state: Partial<Popper.State>) => {
+    if (state.placement) setOurPlacement(state.placement)
+    if (onFirstUpdate) onFirstUpdate(state)
+  }, [onFirstUpdate])
+
+  const [arrowElement, setArrowElement] = React.useState<HTMLDivElement | null>(null)
+  const { styles, attributes, update } = usePopper(reference, divRef.current, {
+    modifiers: [
+      { name: 'arrow', options: { element: arrowElement } },
+      { name: 'offset', options: { offset: [0, 8] } },
+    ],
+    placement,
+    onFirstUpdate: onFirstUpdate_
+  })
+  const update_ = React.useCallback(() => update?.(), [update])
+
+  const logicalDom = React.useContext(LogicalDomContext)
+
+  const popper = <div
+      ref={node => {
+        setDivRef(node)
+        logicalDom.registerDescendant(node)
+      }}
+      style={styles.popper}
+      className='tooltip'
+      {...props}
+      {...attributes.popper}
+    >
+      <TooltipPlacementContext.Provider value={ourPlacement}>
+        {tooltipContent(update_)}
+      </TooltipPlacementContext.Provider>
+      <div ref={setArrowElement}
+        style={styles.arrow}
+        className='tooltip-arrow'
+      />
+    </div>
+
+  // Append the tooltip to the end of document body to avoid layout issues.
+  // (https://github.com/leanprover/vscode-lean4/issues/51)
+  return ReactDOM.createPortal(popper, document.body)
+})
+
+/** A `<span>` element which gets highlighted when hovered over. It is implemented with JS rather
+ * than CSS in order to allow nesting of these elements. When nested, only the smallest nested
+ * element is highlighted. */
+export const HighlightOnHoverSpan = forwardAndUseRef<HTMLSpanElement, React.HTMLProps<HTMLSpanElement>>((props, ref, setRef) => {
+  const [isPointerOver, setIsPointerOver] = React.useState<boolean>(false)
+  const onPointerOver = (e: React.PointerEvent<HTMLSpanElement>) => {
+    if (ref.current && e.target === ref.current)
```

Comment:
Will this break if we add additional elements (such as spans for syntax highlighting)?

Reply:
It will break in some cases with more deeply nested children, but this is intentional. In the case you mention this is what we want, since we only want the innermost span to have a highlight. Note also that React propagates `onPointerOver` events from children in the React tree even if they are not children in the DOM, so hovering over a tooltip will bubble up an event up to the hoverable span that generated the tolltip. Without the `=== ref.current` check this would highlight the span which is not what we want. If we want to make something other than the directly nested text or span highlightable, I think the only reliable way to achieve it is by passing an explicit ref.

Reply:
more comments would help in the code too explaining the tricky parts of the implementation, like the info you had about how React propagates onPointerOver events from children in the React tree even if they are not children in the DOM and how the ref.current check works and so on...